The formats are expected in an AudioStreamRangedDescription object

afigegoznaet opened the pull request:
Fill available formats vectors with default format on startup
Changed containers to plain STL ones, since they are only populated on startup.

afigegoznaet committed:
The formats are expected in an AudioStreamRangedDescription object
Fill available formats vectors with default format on startup
Changed containers to plain STL ones, since they are only populated on startup.

gavv commented:
Hi, thanks a lot for PR!

As I can see, it mixes a few sets of changes:

1. The bugfix - switch from AudioStreamBasicDescription to AudioStreamRangedDescription.

2. Replacing deprecated kAudioObjectPropertyElementMaster with kAudioObjectPropertyElementMain.

3. Replacing DoubleBuffer with raw values in some places.

5. Changing various arguments and return types.

I did the following:

1. I've extracted the bugfix into separate commit: 4319814f5e55c20c300f6888d16b9e78d9066cbb. I slightly modified the approach: instead of populating vectors in ctor, we dynamically construct value based on GetPhysicalFormat / GetVirtualFormat. This preserves the behavior that we had before: if the user will overwrite or overload only GetPhysicalFormat, then GetAvailablePhysicalFormats will automatically use updated value. Also I fixed tests.

2. I've extracted master/main change into separate commit b3539776b967b6ab05055f682b4c16c75518bfab. Since this commit breaks compatibility with macOS versions before 12.0, I added a follow-up commit c197b8b1619abac9eca01ca6f8a3820f6bd44dbf to resolve compatibility problems. I also added multiple macOS versions to CI: dc6dddd0e4b464b3ccc24f892db5b301f798dea5.

3. Regarding this: `Changed containers to plain STL ones, since they are only populated on startup.`. That's not actually true, because we still have setters that can modify those containers on fly. E.g. user can call SetAvailablePhysicalFormatsAsync(), and it will modify available formats. Hence, we should drop this part of changes and keep using DoubleBuffer.

4. The idea of replacing argument and return types looks promising, however there are problems with the implementation:

   * we can't return constant references from methods; we have to return a copy because the stored value may be updated concurrently via setters (like SetAvailablePhysicalFormatsAsync)

   * changing argument type from `const std::vector<...> &` to `std::vector<...>` is a good idea - it gives the caller the right to choose between copy and move; however changing it to `std::vector<...> &&` is bad idea - it forces the caller to use move;

   * capturing local values by reference like this: `RequestConfigurationChange([this, &value]() { ...`  creates a bug, because the lambda passed to that method may be executed asynchronously when the reference wont be valid anymore; we should not do it

   However, in general, the idea of having zero-copy setters is great, we just need to implement it properly: replace const refs with passing by value and use std::move inside. I'll add this change soon.

   We can't have zero-copy getters, unfortunately.

Again, thanks for PR! I can confirm that after this change, Audio MIDI Setup and Garage Band started recognizing the device. I also added you to the AUTHORS file https://github.com/gavv/libASPL/blob/main/AUTHORS.

afigegoznaet commented:
Hey, thanks for the reply, I was thinking you've discarded my commit.
I would suggest to replace python generated classes with template classes,
this would also solve some of the issues you've mentioned.
Cheers!

vin., 25 nov. 2022, 15:33 Victor Gaydov ***@***.***> a scris:

> Hi, thanks a lot for PR!
>
> As I can see, it mixes a few sets of changes:
>
>    1.
>
>    The bugfix - switch from AudioStreamBasicDescription to
>    AudioStreamRangedDescription.
>    2.
>
>    Replacing deprecated kAudioObjectPropertyElementMaster with
>    kAudioObjectPropertyElementMain.
>    3.
>
>    Replacing DoubleBuffer with raw values in some places.
>    4.
>
>    Changing various arguments and return types.
>
> I did the following:
>
>    1.
>
>    I've extracted the bugfix into separate commit: 4319814
>    <https://github.com/gavv/libASPL/commit/4319814f5e55c20c300f6888d16b9e78d9066cbb>.
>    I slightly modified the approach: instead of populating vectors in ctor, we
>    dynamically construct value based on GetPhysicalFormat / GetVirtualFormat.
>    This preserves the behavior that we had before: if the user will overwrite
>    or overload only GetPhysicalFormat, then GetAvailablePhysicalFormats will
>    automatically use updated value. Also I fixed tests.
>    2.
>
>    I've extracted master/main change into separate commit b353977
>    <https://github.com/gavv/libASPL/commit/b3539776b967b6ab05055f682b4c16c75518bfab>.
>    Since this commit breaks compatibility with macOS versions before 12.0, I
>    added a follow-up commit c197b8b
>    <https://github.com/gavv/libASPL/commit/c197b8b1619abac9eca01ca6f8a3820f6bd44dbf>
>    to resolve compatibility problems. I also added multiple macOS versions to
>    CI: dc6dddd
>    <https://github.com/gavv/libASPL/commit/dc6dddd0e4b464b3ccc24f892db5b301f798dea5>
>    .
>    3.
>
>    Regarding this: Changed containers to plain STL ones, since they are
>    only populated on startup.. That's not actually true, because we still
>    have setters that can modify those containers on fly. E.g. user can call
>    SetAvailablePhysicalFormatsAsync(), and it will modify available formats.
>    Hence, we should drop this part of changes and keep using DoubleBuffer.
>    4.
>
>    The idea of replacing argument and return types looks promising,
>    however there are problems with the implementation:
>    -
>
>       we can't return constant references from methods; we have to return
>       a copy because the stored value may be updated concurrently via setters
>       (like SetAvailablePhysicalFormatsAsync)
>       -
>
>       changing argument type from const std::vector<...> & to
>       std::vector<...> is a good idea - it gives the caller the right to
>       choose between copy and move; however changing it to std::vector<...>
>       && is bad idea - it forces the caller to use move;
>       -
>
>       capturing local values by reference like this: RequestConfigurationChange([this,
>       &value]() { ... creates a bug, because the lambda passed to that
>       method may be executed asynchronously when the reference wont be valid
>       anymore; we should not do it
>
>    However, in general, the idea of having zero-copy setters is great, we
>    just need to implement it properly: replace const refs with passing by
>    value and use std::move inside. I'll add this change soon.
>
>    We can't have zero-copy getters, unfortunately.
>
> Again, thanks for PR! I can confirm that after this change, Audio MIDI
> Setup and Garage Band started recognizing the device. I also added you to
> the AUTHORS file https://github.com/gavv/libASPL/blob/main/AUTHORS.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gavv/libASPL/pull/2#issuecomment-1327484976>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABUI5BOQNXDDVD3CXTZQR5LWKC53VANCNFSM5ZZWCNAA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

gavv commented:
I've pushed commit based on your idea. It changes const refs to passing by value for moveable types, and makes most setters zero-copy. See here: ef992e8bb4d6986e3b91b47b94d0aab2dd5e07ac
